addrmgr/addrmanager.go

```diff
@@ -1130,6 +1155,39 @@ func (a *AddrManager) GetBestLocalAddress(remoteAddr *wire.NetAddress) *wire.Net
 	return bestAddress
 }
 
+// IsPeerNaValid asserts if the the provided network address is routable
+// and reachable from the peer that suggested it.
+func (a *AddrManager) IsPeerNaValid(na *wire.NetAddress, peer *wire.NetAddress) bool {
+	net := getNetwork(na)
+	state := getReachabilityFrom(na, peer)
+
+	switch net {
```

Comment:
This all seems way more complicated than it should be.  There are only two true cases.  It would much simpler to do:

```Go
	net := getNetwork(localAddr)
	reach := getReachabilityFrom(localAddr, remoteAddr)
	return (net == IPv4Addr && reach == Ipv4) || (net == IPv6Addr &&
		(reach == Ipv6Weak || reach == Ipv6Strong || reach == Teredo))
```